static/src/javascripts/projects/common/modules/commercial/track-ad.js

```diff
@@ -0,0 +1,41 @@
+define([
+    'Promise',
+    'common/utils/mediator'
+], function (
+    Promise,
+    mediator
+) {
+    var promises;
+
+    function TrackAd() {
+        promises = {};
+    }
+
+    TrackAd.prototype.track = function (id) {
+        promises[id] = promises[id] || new Promise(function (resolve, reject) {
+            var onAdLoaded = function (event) {
+                if (event.slot.getSlotElementId() === id) {
+                    unlisten();
+                    resolve(!event.isEmpty);
+                }
+            };
+
+            var onAllAdsLoaded = function () {
+                unlisten();
+                reject(new Error('Slot ' + id + ' was never loaded'));
+            };
+
+            function unlisten() {
+                mediator.off('modules:commercial:dfp:rendered', onAdLoaded);
+                mediator.off('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
+            }
+
+            mediator.on('modules:commercial:dfp:rendered', onAdLoaded);
+            mediator.on('modules:commercial:dfp:alladsrendered', onAllAdsLoaded);
+        });
+
+        return promises[id];
+    };
+
+    return new TrackAd();
```

Comment:
How are you going to reset the cache during testing? The module will always have the same promise map, am I correct? Are you trying to reset the cache between tests? It is tricky - this'd be a lot simpler with a service layer ;-)

I guess you could put a `reset` method on your class. It's a bit awkward, but I'm not sure if there are any better alternatives at present.

The only other way to reset the map between specs might be to make the injector mock the module with itself. I am not sure if this will work, and is quite a bulky solution in any case:

```javascript
injector.require(['tracker'], function () {
    injector.mock('tracker', new arguments[0].constructor());
});
```